x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Case.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.predicate.conditional;
+
+import org.elasticsearch.xpack.sql.expression.Expression;
+import org.elasticsearch.xpack.sql.expression.Expressions;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder;
+import org.elasticsearch.xpack.sql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.sql.type.DataType;
+import org.elasticsearch.xpack.sql.type.DataTypeConversion;
+import org.elasticsearch.xpack.sql.type.DataTypes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringJoiner;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+/**
+ * Implements the CASE WHEN ... THEN ... ELSE ... END expression
+ */
+public class Case extends ConditionalFunction {
+
+    private final List<IfConditional> conditions;
+    private final Expression defaultElse;
+
+    @SuppressWarnings("unchecked")
+    public Case(Source source, List<Expression> expressions) {
+        super(source, expressions);
+        this.conditions = (List<IfConditional>) (List<?>) expressions.subList(0, expressions.size() - 1);
+        this.defaultElse = expressions.get(expressions.size() - 1);
+        setDataType();
+    }
+
+    public List<IfConditional> conditions() {
+        return conditions;
+    }
+
+    public Expression defaultElse() {
+        return defaultElse;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new Case(source(), newChildren);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Case::new, children());
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        DataType expectedResultDataType = null;
```

Comment:
Why not initializing expectedResultDataType directly with `defaultElse().dataType()`?

Reply:
Because we want to match the results of the `THEN <result>` clauses to the 1st non-null result of them.

Reply:
Wouldn't that happen anyway in the following loop (that checks the conditions)?